### PR TITLE
Add `on_move` hook

### PR DESCRIPTION
**Pull request by iamyulong**
## Summary

Refactors proof movement tracking to use `on_move` hook.

Also, removed composed proof soundness check FIXMEs. See reasoning [here](https://github.com/radixdlt/radixdlt-scrypto/pull/1258/commits/251177b7a668d67cbf469cf7a9eec5020af9a1a2).

### CODE REVIEW

File: radix-engine-interface/src/hooks.rs [hunk 1]
```diff
@@ -17,7 +17,23 @@ pub struct OnDropInput {}
...
+    /// TODO: expose generic information but fully-detailed actor?
+    pub is_to_barrier: bool,
+
+    /// True if the destination actor is auth zone, otherwise false.
+    ///
+    /// TODO: expose generic information but fully-detailed actor?
+    pub is_to_auth_zone: bool,
+
+    /// True if the destination actor is self blueprint.
+    pub is_to_self_blueprint: bool,
```
Comment (talekhinezh) [inline radix-engine-interface/src/hooks.rs:35]: A bit of a weird abstraction? Perhaps better would be exposing the `BlueprintId` of the destination actor?

Comment (iamyulong) [reply radix-engine-interface/src/hooks.rs:35]: Yes, will do

File: radix-engine/src/system/system_callback.rs [hunk 2]
```diff
@@ -658,4 +698,55 @@ impl<C: SystemCallbackObject> KernelCallbackObject for SystemConfig<C> {
...
+            TypeInfoSubstate::Object(object_info) => {
+                let mut service = SystemService::new(api);
+                let definition = service.load_blueprint_definition(
+                    object_info.main_blueprint_id.package_address,
+                    &BlueprintVersionKey {
+                        blueprint: object_info.main_blueprint_id.blueprint_name.clone(),
+                        version: BlueprintVersion::default(),
+                    },
+                )?;
+                if let Some(export) = definition.hook_exports.get(&BlueprintHook::OnMove).cloned() {
```
Comment (talekhinezh) [inline radix-engine/src/system/system_callback.rs:725]: Very cool!

File: radix-engine-interface/src/hooks.rs [hunk 3]
```diff
@@ -16,8 +17,21 @@ pub struct OnDropInput {}
 
 pub type OnDropOutput = ();
 
+// TODO: expose generic information, but fully-detailed actor?
```
Comment (talekhinezh) [inline radix-engine-interface/src/hooks.rs:20]: Can we also add a comment to look to remove some of the too-specific fields (like `is_to_barrier`)

File: radix-engine-interface/src/hooks.rs [hunk 4]
```diff
@@ -16,8 +17,21 @@ pub struct OnDropInput {}
...
-pub struct OnMoveInput {}
+pub struct OnMoveInput {
+    /// True if the node moves from caller to callee, otherwise false.
+    pub is_moving_down: bool,
+
+    /// True if the destination actor is a barrier, otherwise false.
+    pub is_to_barrier: bool,
+
+    /// True if the destination actor is auth zone, otherwise false.
+    pub is_to_auth_zone: bool,
```
Comment (talekhinezh) [inline radix-engine-interface/src/hooks.rs:30]: Couldn't this be removed and just use the `destination_blueprint_id`?

**Review by talekhinezh — APPROVED** [review]

File: radix-engine-interface/src/hooks.rs (hunk 3 above)
Comment (iamyulong) [reply radix-engine-interface/src/hooks.rs:20]: Yes

File: radix-engine-interface/src/hooks.rs (hunk 4 above)
Comment (iamyulong) [reply radix-engine-interface/src/hooks.rs:30]: Yes

**Pull request merged**